File: mobile_flutter/ios/Runner/Bridges/TurnaLaunchBridge.swift
import Flutter
import Foundation

final class TurnaLaunchBridge {
  private let launchScheme = "turna"
  private var launchChannel: FlutterMethodChannel?
  private var isBridgeReady = false
  private var pendingLaunchURL: String?

  func configure(binaryMessenger: FlutterBinaryMessenger) {
    guard launchChannel == nil else { return }

    let channel = FlutterMethodChannel(
      name: "turna/launch",
      binaryMessenger: binaryMessenger
    )
    channel.setMethodCallHandler { [weak self] call, result in
      switch call.method {
      case "launchBridgeReady":
        self?.isBridgeReady = true
        TurnaLogger.debug(
          "launch",
          "bridge ready",
          details: ["hasURL": (self?.pendingLaunchURL != nil)]
        )
        self?.dispatchPendingLaunchURLIfReady()
        result(nil)
      case "consumeInitialUrl":
        let url = self?.pendingLaunchURL
        self?.pendingLaunchURL = nil
        TurnaLogger.debug(
          "launch",
          "consume initial url",
          details: ["hasURL": (url != nil)]
        )
        result(url)
      default:
        result(FlutterMethodNotImplemented)
      }
    }
    launchChannel = channel
    dispatchPendingLaunchURLIfReady()
  }

  @discardableResult
  func handleIncomingURL(_ url: URL) -> Bool {
    guard url.scheme?.lowercased() == launchScheme else {
      return false
    }
    pendingLaunchURL = url.absoluteString
    TurnaLogger.debug(
      "launch",
      "incoming url",
      details: ["url": sanitizedURLString(url)]
    )
    dispatchPendingLaunchURLIfReady()
    return true
  }

  private func dispatchPendingLaunchURLIfReady() {
    guard let url = pendingLaunchURL else {
      return
    }
    guard isBridgeReady else {
      TurnaLogger.debug("launch", "dispatch postponed", details: ["reason": "bridge_not_ready"])
      return
    }
    guard launchChannel != nil else {
      TurnaLogger.debug("launch", "dispatch postponed", details: ["reason": "channel_missing"])
      return
    }
    pendingLaunchURL = nil
    TurnaLogger.debug(
      "launch",
      "dispatching url to flutter",
      details: ["url": sanitizedURLString(URL(string: url) ?? URL(fileURLWithPath: url))]
    )
    launchChannel?.invokeMethod("launchUrlUpdated", arguments: url)
  }

  private func sanitizedURLString(_ url: URL) -> String {
    guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
      return url.absoluteString
    }
    if let items = components.queryItems, !items.isEmpty {
      components.queryItems = items.map { item in
        URLQueryItem(name: item.name, value: item.value == nil ? nil : "redacted")
      }
    }
    return components.string ?? url.absoluteString
  }
}

File: mobile_flutter/android/app/src/main/kotlin/com/turna/chat/bridges/TurnaLaunchBridge.kt
package com.turna.chat.bridges

import android.content.Intent
import android.net.Uri
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import java.util.Locale

class TurnaLaunchBridge {
    private var launchChannel: MethodChannel? = null
    private var isBridgeReady = false
    private var pendingLaunchUrl: String? = null

    fun configure(binaryMessenger: BinaryMessenger) {
        if (launchChannel != null) return
        val channel = MethodChannel(binaryMessenger, "turna/launch")
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "launchBridgeReady" -> {
                    isBridgeReady = true
                    TurnaLogger.debug(
                        "launch",
                        "bridge ready",
                        mapOf("hasUrl" to (pendingLaunchUrl != null)),
                    )
                    dispatchPendingLaunchUrlIfReady()
                    result.success(null)
                }

                "consumeInitialUrl" -> {
                    val url = pendingLaunchUrl
                    pendingLaunchUrl = null
                    TurnaLogger.debug(
                        "launch",
                        "consume initial url",
                        mapOf("hasUrl" to (url != null)),
                    )
                    result.success(url)
                }

                else -> result.notImplemented()
            }
        }
        launchChannel = channel
        dispatchPendingLaunchUrlIfReady()
    }

    fun captureIncomingIntent(
        intent: Intent?,
        notifyFlutter: Boolean,
    ): Boolean {
        if (intent == null || intent.action != Intent.ACTION_VIEW) {
            return false
        }
        val data = intent.data ?: return false
        if (data.scheme?.lowercase(Locale.US) != "turna") {
            return false
        }
        pendingLaunchUrl = data.toString()
        TurnaLogger.debug(
            "launch",
            "incoming url",
            mapOf("url" to sanitizeUrl(data)),
        )
        if (notifyFlutter) {
            dispatchPendingLaunchUrlIfReady()
        }
        return true
    }

    private fun dispatchPendingLaunchUrlIfReady() {
        val url = pendingLaunchUrl ?: return
        val channel = launchChannel
        if (!isBridgeReady) {
            TurnaLogger.debug("launch", "dispatch postponed", mapOf("reason" to "bridge_not_ready"))
            return
        }
        if (channel == null) {
            TurnaLogger.debug("launch", "dispatch postponed", mapOf("reason" to "channel_missing"))
            return
        }
        pendingLaunchUrl = null
        TurnaLogger.debug(
            "launch",
            "dispatching url to flutter",
            mapOf("url" to sanitizeUrl(Uri.parse(url))),
        )
        channel.invokeMethod("launchUrlUpdated", url)
    }

    private fun sanitizeUrl(uri: Uri): String {
        val query =
            if (uri.queryParameterNames.isEmpty()) {
                null
            } else {
                uri.queryParameterNames.joinToString("&") { "$it=redacted" }
            }
        val base =
            buildString {
                append(uri.scheme ?: "")
                append("://")
                if (!uri.authority.isNullOrBlank()) {
                    append(uri.authority)
                }
                append(uri.path.orEmpty())
            }
        return if (query == null) base else "$base?$query"
    }
}
